osutil/cmp_test.go

```diff
@@ -96,6 +94,51 @@ func (ts *CmpTestSuite) TestCmpStreams(c *C) {
 		{"hello", "world", false},
 		{"hello", "hell", false},
 	} {
-		c.Assert(StreamsEqual(strings.NewReader(x.a), strings.NewReader(x.b)), Equals, x.r)
+		c.Assert(osutil.StreamsEqual(strings.NewReader(x.a), strings.NewReader(x.b)), Equals, x.r)
+	}
+}
+
+func (s *CmpTestSuite) TestStreamsEqualChunked(c *C) {
+	text := "marry had a little lamb"
+
+	// Passing the same stream twice is not mishandled.
+	readerA := bytes.NewReader([]byte(text))
+	readerB := readerA
+	eq := osutil.StreamsEqualChunked(readerA, readerB, 0)
+	c.Check(eq, Equals, true)
+
+	// Passing two streams with the same content works as expected. Note that
+	// we are using different block sizes to check for additional edge cases.
+	for _, chunkSize := range []int{0, 1, len(text) / 2, len(text), len(text) + 1} {
+		readerA = bytes.NewReader([]byte(text))
+		readerB = bytes.NewReader([]byte(text))
+		eq := osutil.StreamsEqualChunked(readerA, readerB, chunkSize)
+		c.Check(eq, Equals, true, Commentf("chunk size %d", chunkSize))
+	}
+
+	// Passing two streams with unequal contents but equal length works as
+	// expected.
+	for _, chunkSize := range []int{0, 1, len(text) / 2, len(text), len(text) + 1} {
+		comment := Commentf("chunk size %d", chunkSize)
+		readerA = bytes.NewReader([]byte(strings.ToLower(text)))
+		readerB = bytes.NewReader([]byte(strings.ToUpper(text)))
+		eq = osutil.StreamsEqualChunked(readerA, readerB, chunkSize)
+		c.Check(eq, Equals, false, comment)
+	}
+
+	// Passing two streams with different length works as expected.
+	// Note that this is not used by EnsureDirState in practice.
+	for _, chunkSize := range []int{0, 1, len(text) / 2, len(text), len(text) + 1} {
+		comment := Commentf("A: %q, B: %q, chunk size %d", text, text[:len(text)/2], chunkSize)
+		readerA = bytes.NewReader([]byte(text))
+		readerB = bytes.NewReader([]byte(text[:len(text)/2]))
+		eq = osutil.StreamsEqualChunked(readerA, readerB, chunkSize)
+		c.Check(eq, Equals, false, comment)
+
+		// Readers passed the other way around.
+		readerA = bytes.NewReader([]byte(text))
+		readerB = bytes.NewReader([]byte(text[:len(text)/2]))
+		eq = osutil.StreamsEqualChunked(readerB, readerA, chunkSize)
+		c.Check(eq, Equals, false, comment)
```

Comment:
I think we're missing a test case for a scenario in which the string lengths are equal, and they differ at the end, where the different part is < read chunk size.

eg.
```
one := bytes.NewBufferString("mary")
two := bytes.NewBufferString("marY")
// both readers give UnexpectedEOF on the last read
eq := osutil.StreamsEqualChunked(one, two, 3)
c.Check(eq, Equals, false)
```

Reply:
Done now